tik/lib/CartographerKernel.cpp

```diff
@@ -291,500 +281,926 @@ namespace TraceAtlas::tik
+                                    for (auto succ : successors(phi->getIncomingBlock(j)))
+                                    {
+                                        if (succ != phi->getIncomingBlock(j) && Entrances.find(GetBlockID(succ)) != Entrances.end() && scopedBlocks.find(phi->getIncomingBlock(j)) == scopedBlocks.end())
+                                        {
+                                            import = true;
+                                            auto sExtVal = GetValueID(operand);
+                                            if (kernelIE.find(sExtVal) == kernelIE.end())
+                                            {
+                                                KernelImports.push_back(sExtVal);
+                                                kernelIE.insert(sExtVal);
+                                            }
+                                        }
+                                    }
+                                    // below captures values who come from far away places and enter through the front door
+                                    if (!import)
                                     {
-                                        KernelImports.push_back(sExtVal);
+                                        if (scopedBlocks.find(operand->getParent()) == scopedBlocks.end() && (Entrances.find(GetBlockID(phi->getIncomingBlock(j))) != Entrances.end() || scopedBlocks.find(phi->getIncomingBlock(j)) != scopedBlocks.end()))
+                                        {
+                                            auto sExtVal = GetValueID(operand);
+                                            if (kernelIE.find(sExtVal) == kernelIE.end())
+                                            {
+                                                KernelImports.push_back(sExtVal);
+                                                kernelIE.insert(sExtVal);
+                                            }
+                                        }
                                     }
                                 }
                             }
                         }
-                        else
+                        else if (scopedBlocks.find(operand->getParent()) == scopedBlocks.end())
                         {
-                            if (find(KernelImports.begin(), KernelImports.end(), GetValueID(ar)) == KernelImports.end())
+                            if (scopedFuncs.find(operand->getParent()->getParent()) == scopedFuncs.end())
                             {
-                                KernelImports.push_back(GetValueID(ar));
+                                auto sExtVal = GetValueID(operand);
+                                if (kernelIE.find(sExtVal) == kernelIE.end())
+                                {
+                                    KernelImports.push_back(sExtVal);
+                                    kernelIE.insert(sExtVal);
+                                }
```

Comment:
You no longer check for arguments. Is there a reason you can ignore them?

Reply:
That check is now further up, right after the error handling if() block